File: app/src/main/java/localhost/android/me/database/repo/ProviderRepo.java
package localhost.android.me.database.repo;

import android.content.Context;
import java.util.List;
import localhost.android.me.database.dao.ProviderDao;
import localhost.android.me.database.MarketEnablerDatabase;
import localhost.android.me.database.entity.Provider;
import localhost.android.me.database.model.ProviderWithCountry;

public class ProviderRepo
{
    private ProviderDao providerDao;

    public ProviderRepo(Context context)
    {
        MarketEnablerDatabase db = MarketEnablerDatabase.getDatabase(context);
        this.providerDao = db.providerDao();
    }

    public List<ProviderWithCountry> getProviders() throws InterruptedException
    {
        GetAll getAll = new GetAll(this.providerDao);
        Thread thread = new Thread(getAll);
        thread.start();
        thread.join();
        return getAll.list();
    }

    public void insertProvider(Provider provider) throws InterruptedException
    {
        Insert insert = new Insert(this.providerDao, provider);
        Thread thread = new Thread(insert);
        thread.start();
        thread.join();
    }

    public void updateProvider(Provider provider) throws InterruptedException
    {
        Update update = new Update(this.providerDao, provider);
        Thread thread = new Thread(update);
        thread.start();
        thread.join();
    }

    public void deleteProvider(Provider provider) throws InterruptedException
    {
        Delete delete = new Delete(this.providerDao, provider);
        Thread thread = new Thread(delete);
        thread.start();
        thread.join();
    }

    public class GetAll implements Runnable
    {
        private ProviderDao providerDao;
        private volatile List<ProviderWithCountry> providers;

        GetAll(ProviderDao providerDao)
        {
            this.providerDao = providerDao;
        }

        @Override
        public void run()
        {
            this.providers = this.providerDao.getAll();
        }

        private List<ProviderWithCountry> list()
        {
            return this.providers;
        }
    }

    public class Insert implements Runnable
    {
        private ProviderDao providerDao;
        private Provider provider;

        Insert(ProviderDao providerDao, Provider provider)
        {
            this.providerDao = providerDao;
            this.provider = provider;
        }

        @Override
        public void run()
        {
            this.providerDao.insert(this.provider);
        }
    }

    public class Update implements Runnable
    {
        private ProviderDao providerDao;
        private Provider provider;

        Update(ProviderDao providerDao, Provider provider)
        {
            this.providerDao = providerDao;
            this.provider = provider;
        }

        @Override
        public void run()
        {
            this.providerDao.update(this.provider);
        }
    }

    public class Delete implements Runnable
    {
        private ProviderDao providerDao;
        private Provider provider;

        Delete(ProviderDao providerDao, Provider provider)
        {
            this.providerDao = providerDao;
            this.provider = provider;
        }

        @Override
        public void run()
        {
            this.providerDao.delete(this.provider);
        }
    }
}

File: app/src/main/java/localhost/android/me/database/repo/ProviderRepo.kt
package localhost.android.me.database.repo

import android.content.Context
import android.os.AsyncTask
import localhost.android.me.database.Callback
import localhost.android.me.database.dao.ProviderDao
import localhost.android.me.database.MarketEnablerDatabase
import localhost.android.me.database.entity.Provider
import localhost.android.me.database.model.ProviderWithCountry

class ProviderRepo(private var context: Context)
{
    private var providerDao: ProviderDao

    init {
        val db = MarketEnablerDatabase.getDatabase(this.context)
        this.providerDao = db.providerDao()
    }

    fun getProviders(resultCallback: Callback.OnProviderResultListener)
    {
        GetAllTask(this.providerDao, resultCallback).execute()
    }

    fun insertProvider(provider: Provider, doneCallback: Callback.OnDoneListener)
    {
        InsertTask(this.providerDao, provider, doneCallback).execute()
    }

    fun updateProvider(provider: Provider, doneCallback: Callback.OnDoneListener)
    {
        UpdateTask(this.providerDao, provider, doneCallback).execute()
    }

    fun deleteProvider(provider: Provider)
    {
        DeleteTask(this.providerDao, provider).execute()
    }

    class GetAllTask(
        private var providerDao: ProviderDao,
        private var resultCallback: Callback.OnProviderResultListener
    ) : AsyncTask<Void?, Void, List<ProviderWithCountry>>()
    {
        override fun doInBackground(vararg p0: Void?): List<ProviderWithCountry>
        {
            return this.providerDao.getAll()
        }

        override fun onPostExecute(providers: List<ProviderWithCountry>)
        {
            this.resultCallback.onResult(providers)
        }
    }

    class InsertTask(
        private var providerDao: ProviderDao,
        private var provider: Provider,
        private var doneCallback: Callback.OnDoneListener
    ) : AsyncTask<Void?, Void?, Void?>()
    {
        override fun doInBackground(vararg p0: Void?): Void?
        {
            this.providerDao.insert(this.provider)
            return null
        }

        override fun onPostExecute(result: Void?)
        {
            this.doneCallback.onSuccess()
        }
    }

    class UpdateTask(
        private var providerDao: ProviderDao,
        private var provider: Provider,
        private var doneCallback: Callback.OnDoneListener
    ) : AsyncTask<Void?, Void?, Void?>()
    {
        override fun doInBackground(vararg p0: Void?): Void?
        {
            this.providerDao.update(this.provider)
            return null
        }

        override fun onPostExecute(result: Void?)
        {
            this.doneCallback.onSuccess()
        }
    }

    class DeleteTask(
        private var providerDao: ProviderDao,
        private var provider: Provider
    ) : AsyncTask<Void?, Void?, Void?>()
    {
        override fun doInBackground(vararg p0: Void?): Void?
        {
            this.providerDao.delete(this.provider)
            return null
        }
    }
}
